File: commands/gacha.ts
/* eslint-disable no-unreachable */
import path from 'path';
import {
  EmbedBuilder, ActionRowBuilder, ButtonBuilder, ButtonStyle,
} from 'discord.js';
import { Command } from './classes/Command';

let pools: { namesData: any; characterPool: any[]; lightconePool: any[] } | null = null;

async function getPools() {
  if (!pools) {
    const [charactersRaw, lightconesRaw, hsrNames] = await Promise.all([
      Bun.file(path.join(__dirname, '../data/hsrCharacters.json')).json(),
      Bun.file(path.join(__dirname, '../data/hsrLC.json')).json(),
      Bun.file(path.join(__dirname, '../data/hsr.json')).json(),
    ]);
    pools = {
      namesData: hsrNames,
      characterPool: Object.values(charactersRaw),
      lightconePool: Object.values(lightconesRaw),
    };
  }
  return pools!;
}

class Gacha extends Command {
  namesData: any;
  characterPool: any[];
  lightconePool: any[];

  constructor(client: any) {
    super(client, 'gacha', 'TECHNICAL TEST, WORK IN PROGRESS', [
      {
        name: 'amount',
        description: 'Number of rolls (1 or 10)',
        type: 4,
        required: true,
        choices: [
          { name: '1', value: 1 },
          { name: '10', value: 10 },
        ],
      },
    ], { blame: 'xei' });

    this.namesData = null;
    this.characterPool = [];
    this.lightconePool = [];
  }

  getRandomItem(pool: any[]) {
    return pool.length > 0 ? pool[Math.floor(Math.random() * pool.length)] : null;
  }

  calculatePityRate(pityCount: number) {
    return pityCount >= 74 ? Math.min(1, (pityCount - 73) * 0.1) : 0.006;
  }

  getItemDetails(item: any) {
    if (!item) return { name: 'Unknown', imagePath: null, rarity: 'Unknown' };

    const nameHash = item.AvatarName?.Hash?.toString() || item.EquipmentName?.Hash?.toString();
    const name = this.namesData.en[nameHash] || 'Unknown';
    const imagePath = item.AvatarCutinFrontImgPath || item.ImagePath || null;
    const rarity = item.Rarity || 'Unknown';

    return {
      name,
      imageUrl: imagePath ? `https://enka.network${imagePath}` : null,
      rarity,
    };
  }

  async run(interaction: any): Promise<void> {
    await interaction.editReply({ content: 'gacha is not ready yet', ephemeral: true });
    return;

    const loaded = await getPools();
    this.namesData = loaded.namesData;
    this.characterPool = loaded.characterPool;
    this.lightconePool = loaded.lightconePool;

    const amount = interaction.options.getInteger('amount');
    const userId = interaction.user.id;
    let pityCount = await this.client.db.user.getUserAttr(userId, 'pity');
    const dinonuggies = await this.client.db.user.getUserAttr(userId, 'dinonuggies');

    const costPerRoll = 160;
    const totalCost = costPerRoll * amount;

    if (dinonuggies < totalCost) {
      await interaction.editReply({
        embeds: [new EmbedBuilder()
          .setTitle('Not enough dinonuggies!')
          .setDescription(`You need ${totalCost}, but you only have ${dinonuggies}.`)
          .setColor(0xFF0000)],
      });
      return;
    }

    await this.client.db.user.addUserAttr(userId, 'dinonuggies', -totalCost);

    const results: any[] = [];
    let gotFiveStar = false;

    for (let i = 0; i < amount; i += 1) {
      let rollResult;
      const pityRate = this.calculatePityRate(pityCount);
      const roll = Math.random();

      if (roll < pityRate) {
        rollResult = Math.random() < 0.5
          ? this.getRandomItem(this.characterPool.filter((c) => c.Rarity === 5))
          : this.getRandomItem(this.lightconePool.filter((lc) => lc.Rarity === 5));
        gotFiveStar = true;
        pityCount = 0;
      } else if (roll < 0.056) {
        rollResult = Math.random() < 0.5
          ? this.getRandomItem(this.characterPool.filter((c) => c.Rarity === 4))
          : this.getRandomItem(this.lightconePool.filter((lc) => lc.Rarity === 4));
        pityCount += 1;
      } else {
        rollResult = this.getRandomItem(this.lightconePool.filter((lc) => lc.Rarity === 3));
        pityCount += 1;
      }

      const itemDetails = this.getItemDetails(rollResult);
      results.push(itemDetails);

      const itemType = this.characterPool.some((c) => c.name === itemDetails.name) ? 'Character' : 'Lightcone';
      this.client.db.gacha.addGachaItem(userId, itemDetails.name, itemType, itemDetails.rarity);
    }

    await this.client.db.user.setUserAttr(userId, 'pity', pityCount);

    let currentIndex = 0;

    const updateMessage = async (i: any) => {
      const item = results[currentIndex];
      const embed = new EmbedBuilder()
        .setTitle(`Gacha Roll #${currentIndex + 1}`)
        .setDescription(`**${item.name}**`)
        .setImage(item.imageUrl)
        .setColor(gotFiveStar ? 0xFFD700 : 0x00FF00)
        .setFooter({ text: `Pity: ${pityCount}` });

      const row = new ActionRowBuilder()
        .addComponents(
          new ButtonBuilder()
            .setCustomId('nextRoll')
            .setLabel('➡️ Next')
            .setStyle(ButtonStyle.Primary)
            .setDisabled(currentIndex === results.length - 1),
          new ButtonBuilder()
            .setCustomId('skipResults')
            .setLabel('Skip to Results')
            .setStyle(ButtonStyle.Danger),
        );

      await i.update({ embeds: [embed], components: [row] });
    };

    const initialEmbed = new EmbedBuilder()
      .setTitle('Gacha Roll #1')
      .setDescription(`**${results[0].name}**`)
      .setImage(results[0].imageUrl)
      .setColor(gotFiveStar ? 0xFFD700 : 0x00FF00)
      .setFooter({ text: `Pity: ${pityCount}` });

    const initialRow = new ActionRowBuilder()
      .addComponents(
        new ButtonBuilder()
          .setCustomId('nextRoll')
          .setLabel('➡️ Next')
          .setStyle(ButtonStyle.Primary),
        new ButtonBuilder()
          .setCustomId('skipResults')
          .setLabel('Skip to Results')
          .setStyle(ButtonStyle.Danger),
      );

    const message = await interaction.editReply({ embeds: [initialEmbed], components: [initialRow] });

    const collector = message.createMessageComponentCollector({ time: 60000 });

    collector.on('collect', async (i: any) => {
      if (i.customId === 'next_roll') {
        currentIndex += 1;
        if (currentIndex < results.length) {
          await updateMessage(i);
        }
      } else if (i.customId === 'skip_results') {
        collector.stop();
      }
    });

    collector.on('end', async () => {
      await message.edit({ components: [] });
      const finalEmbed = new EmbedBuilder()
        .setTitle(`Gacha Roll Results for ${interaction.user.username} (${amount})`)
        .setDescription(results.map((item) => `**${item.name}** - ${item.rarity}★`).join('\n'))
        .setColor(gotFiveStar ? 0xFFD700 : 0x00FF00);
      await interaction.followUp({ embeds: [finalEmbed] });
    });
  }
}

export default Gacha;

File: src/commands/gacha.rs
use crate::{Context, Error};
use poise::serenity_prelude as serenity;
use futures::StreamExt;
use rand::Rng;

/// TECHNICAL TEST, WORK IN PROGRESS
#[poise::command(slash_command)]
pub async fn gacha(
    ctx: Context<'_>,
    #[description = "Number of rolls (1 or 10)"]
    #[choices(1, 10)]
    amount: i32,
) -> Result<(), Error> {
    ctx.defer().await?;

    let user_id = ctx.author().id.to_string();
    let user = ctx.data().db.get_user(&user_id).await?;
    let mut pity_count = user.pity;
    let dinonuggies = user.dinonuggies;

    let cost_per_roll = 160.0;
    let total_cost = cost_per_roll * amount as f64;

    if dinonuggies < total_cost {
        ctx.send(poise::CreateReply::default().embed(
            serenity::CreateEmbed::new()
                .title("Not enough dinonuggies!")
                .description(format!("You need {}, but you only have {}.", total_cost, dinonuggies))
                .color(0xFF0000)
        )).await?;
        return Ok(());
    }

    // Deduct cost
    ctx.data().db.add_user_attr(&user_id, "dinonuggies", -total_cost).await?;

    let mut results = Vec::new();
    let mut got_five_star = false;

    {
        let mut rng = rand::thread_rng();

        for _ in 0..amount {
            let pity_rate = if pity_count >= 74 {
                ((pity_count - 73) as f64 * 0.1).min(1.0)
            } else {
                0.006
            };

            let roll: f64 = rng.gen();
            let roll_result;

            if roll < pity_rate {
                // 5 star
                let pool = if rng.gen_bool(0.5) {
                    ctx.data().hsr_characters.as_object().unwrap().values()
                        .filter(|v| v["Rarity"].as_i64() == Some(5))
                        .collect::<Vec<_>>()
                } else {
                    ctx.data().hsr_lc.as_object().unwrap().values()
                        .filter(|v| v["Rarity"].as_i64() == Some(5))
                        .collect::<Vec<_>>()
                };
                roll_result = pool[rng.gen_range(0..pool.len())].clone();
                got_five_star = true;
                pity_count = 0;
            } else if roll < 0.056 {
                // 4 star
                let pool = if rng.gen_bool(0.5) {
                    ctx.data().hsr_characters.as_object().unwrap().values()
                        .filter(|v| v["Rarity"].as_i64() == Some(4))
                        .collect::<Vec<_>>()
                } else {
                    ctx.data().hsr_lc.as_object().unwrap().values()
                        .filter(|v| v["Rarity"].as_i64() == Some(4))
                        .collect::<Vec<_>>()
                };
                roll_result = pool[rng.gen_range(0..pool.len())].clone();
                pity_count += 1;
            } else {
                // 3 star
                let pool: Vec<_> = ctx.data().hsr_lc.as_object().unwrap().values()
                    .filter(|v| v["Rarity"].as_i64() == Some(3))
                    .collect();
                roll_result = pool[rng.gen_range(0..pool.len())].clone();
                pity_count += 1;
            }

            let name_hash = roll_result["AvatarName"]["Hash"].as_str()
                .or_else(|| roll_result["EquipmentName"]["Hash"].as_str())
                .unwrap_or_default();
            let name = ctx.data().hsr_names["en"][name_hash].as_str().unwrap_or("Unknown").to_string();
            let image_path = roll_result["AvatarCutinFrontImgPath"].as_str()
                .or_else(|| roll_result["ImagePath"].as_str());
            let rarity = roll_result["Rarity"].as_i64().unwrap_or(0);

            results.push(GachaResult {
                name,
                image_url: image_path.map(|p| format!("https://enka.network{}", p)),
                rarity,
            });
        }
    }

    // Update pity
    ctx.data().db.set_user_attr(&user_id, "pity", pity_count as f64).await?;

    let mut current_index = 0;

    let generate_embed = |index: usize, results: &[GachaResult], pity: i64, got_5: bool| {
        let item = &results[index];
        let mut embed = serenity::CreateEmbed::new()
            .title(format!("Gacha Roll #{}", index + 1))
            .description(format!("**{}**", item.name))
            .color(if got_5 { 0xFFD700 } else { 0x00FF00 })
            .footer(serenity::CreateEmbedFooter::new(format!("Pity: {}", pity)));
        
        if let Some(url) = &item.image_url {
            embed = embed.image(url);
        }
        embed
    };

    let ctx_id = ctx.id();
    let next_id = format!("{}next", ctx_id);
    let skip_id = format!("{}skip", ctx_id);

    let reply = ctx.send(poise::CreateReply::default()
        .embed(generate_embed(current_index, &results, pity_count, got_five_star))
        .components(vec![serenity::CreateActionRow::Buttons(vec![
            serenity::CreateButton::new(&next_id).label("➡️ Next").style(serenity::ButtonStyle::Primary).disabled(results.len() == 1),
            serenity::CreateButton::new(&skip_id).label("Skip to Results").style(serenity::ButtonStyle::Danger),
        ])])
    ).await?;

    let mut interaction_stream = reply
        .message()
        .await?
        .await_component_interactions(ctx.serenity_context())
        .timeout(std::time::Duration::from_secs(60))
        .stream();

    while let Some(mci) = interaction_stream.next().await {
        if mci.data.custom_id == next_id {
            current_index += 1;
            if current_index < results.len() {
                mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::UpdateMessage(
                    serenity::CreateInteractionResponseMessage::new()
                        .embed(generate_embed(current_index, &results, pity_count, got_five_star))
                        .components(vec![serenity::CreateActionRow::Buttons(vec![
                            serenity::CreateButton::new(&next_id).label("➡️ Next").style(serenity::ButtonStyle::Primary).disabled(current_index == results.len() - 1),
                            serenity::CreateButton::new(&skip_id).label("Skip to Results").style(serenity::ButtonStyle::Danger),
                        ])])
                )).await?;
            }
        } else if mci.data.custom_id == skip_id {
            break;
        }
    }

    // Final results
    let final_embed = serenity::CreateEmbed::new()
        .title(format!("Gacha Roll Results for {} ({})", ctx.author().name, amount))
        .description(results.iter().map(|item| format!("**{}** - {}★", item.name, item.rarity)).collect::<Vec<_>>().join("\n"))
        .color(if got_five_star { 0xFFD700 } else { 0x00FF00 });

    ctx.send(poise::CreateReply::default()
        .embed(final_embed)
        .components(vec![])
    ).await?;

    Ok(())
}

struct GachaResult {
    name: String,
    image_url: Option<String>,
    rarity: i64,
}
